presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java

```diff
@@ -95,6 +100,84 @@ public Optional<PlanNode> visitPlan(PlanNode node, Context context)
         return Optional.empty();
     }
 
+    @Override
+    public Optional<PlanNode> visitJoin(JoinNode node, Context context)
+    {
+        if (strategy == DEFAULT) {
+            return Optional.empty();
+        }
+        if (node.getType().equals(JoinNode.Type.RIGHT)) {
+            return visitJoin(node.flipChildren(), context);
+        }
+        List<PlanNode> sources = new ArrayList<>();
+        ImmutableList.Builder<RowExpression> allFilters = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> criterias = ImmutableList.builder();
+        Stack<JoinNode> stack = new Stack<>();
+
+        stack.push(node);
+        while (!stack.empty()) {
+            JoinNode top = stack.pop();
+            top.getCriteria().forEach(criterias::add);
+            // ReorderJoins can move predicates between `criteria` and `filters`, so we put all equalities
+            // in `criteria` to make it consistent.
+            if (top.getFilter().isPresent()) {
+                List<RowExpression> filters = extractConjuncts(top.getFilter().get());
+                filters.forEach(filter -> {
+                    Optional<JoinNode.EquiJoinClause> criteria = toEquiJoinClause(filter);
+                    criteria.ifPresent(criterias::add);
+                    allFilters.add(filter);
+                });
+            }
+            top.getSources().forEach(source -> {
+                if (source instanceof JoinNode
+                        && ((JoinNode) source).getType().equals(node.getType())
+                        && shouldMergeJoinNodes(node.getType())) {
+                    stack.push((JoinNode) source);
+                }
+                else {
+                    sources.add(source);
+                }
+            });
+        }
+
+        // Sort sources if all are INNER, or full outer join of 2 nodes
+        if (shouldMergeJoinNodes(node.getType()) || (node.getType().equals(JoinNode.Type.FULL) && sources.size() == 2)) {
+            Map<PlanNode, String> sourceKeys = new IdentityHashMap<>();
+            for (PlanNode source : sources) {
+                Optional<CanonicalPlan> canonicalSource = generateCanonicalPlan(source, strategy, objectMapper);
+                if (!canonicalSource.isPresent()) {
```

Comment:
what does it mean if canonicalSource is  absent?  (and maybe add a comment with the answer)

Reply:
Added comment on generateCanonicalPlan instead, as some other parts use it as well